### PR TITLE
openmpi: make fortran support optional, disable fortran on aarch64-darwin

### LINKED ISSUE #127608: openmpi: build fails on aarch64-darwin
**Describe the bug**

I'm on a mission to get `gdal` (and thence `postgis`) compiling on M1 Macs, and having fixed `hdf4`, the next stumbling block is `openmpi` 4.1.1. Sadly, the build fails with `checking size of Fortran CHARACTER... configure: WARNING: Could not determine size of CHARACTER`, presumably related to `gfortran`.

**To Reproduce**
1. `nix-build --argstr localSystem aarch64-darwin -A openmpi`

**Expected behavior**

**Screenshots**

```
checking to see if Fortran compilers need additional linker flags... none
checking  external symbol convention... single underscore
checking if C and Fortran are link compatible... yes
checking to see if Fortran compiler likes the C++ exception flags... skipped (no C++ exceptions flags)
checking to see if mpifort compiler needs additional linker flags... -Wl,-commons,use_dylibs
checking if Fortran compiler supports CHARACTER... yes
checking size of Fortran CHARACTER... configure: WARNING: Could not determine size of CHARACTER
configure: WARNING: See config.log for details
configure: error: Cannot continue
builder for '/nix/store/rnm0zh3klzh8i5n7llppncpa2rr0ifwp-openmpi-4.1.1.drv' failed with exit code 1
cannot build derivation '/nix/store/94l8ii0kr5pl9l9ahirwwvz2szq2m20p-netcdf-4.7.4.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/kn1kaph2gkqd6rc1769mqv1i49r3kk67-gdal-3.2.2.drv': 1 dependencies couldn't be built
error: build of '/nix/store/kn1kaph2gkqd6rc1769mqv1i49r3kk67-gdal-3.2.2.drv' failed
```

**Additional context**

This works fine when built for x86_64 on the same machine.

**Notify maintainers**

**Metadata**

 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.5.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.10`
 - channels(steve): `"darwin, nixpkgs-21.11pre295944.0747387223e"`
 - nixpkgs: `/Users/steve/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - openmpi
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

- make fortran support optional
- disable fortran on aarch64-darwin
- fixes https://github.com/NixOS/nixpkgs/issues/127608

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by markuskowa** [comment]
I am in principle not against making the Fortran support optional. Is there no option to rather fix gortran on aarch64-darwin? That would be a more natural first step. Lot's of packages need will need the Fortran support.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 127694` at 33d3f270 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages marked as broken and skipped:</summary>
  <ul>
    <li>grass</li>
    <li>python38Packages.dask-gateway</li>
    <li>python38Packages.dask-jobqueue</li>
    <li>python38Packages.dask-ml</li>
    <li>python38Packages.dask-mpi</li>
    <li>python38Packages.dask-xgboost</li>
    <li>python38Packages.datashader</li>
    <li>python38Packages.fipy</li>
    <li>python38Packages.pygmt</li>
    <li>python38Packages.streamz</li>
    <li>python38Packages.stumpy</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python39Packages.dask-gateway</li>
    <li>python39Packages.dask-jobqueue</li>
    <li>python39Packages.dask-mpi</li>
    <li>python39Packages.dask-xgboost</li>
    <li>python39Packages.fipy</li>
    <li>python39Packages.pygmt</li>
    <li>python39Packages.streamz</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>qlandkartegt</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/d30c8d9052673e904d119e7e1bf64f31">python38Packages.fenics</a></li>
    <li><a href="https://gist.github.com/b9722c20d50ad6b7323a16b55bb3cfad">python39Packages.fenics</a></li>
  </ul>
</details>
<details>
  <summary>83 packages skipped due to time constraints:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>cdo</li>
    <li>cloudcompare</li>
    <li>elmerfem</li>
    <li>ezminc</li>
    <li>freecad</li>
    <li>gdal (python38Packages.gdal)</li>
    <li>gdal_2</li>
    <li>globalarrays</li>
    <li>gplates</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>40 packages built successfully:</summary>
  <ul>
    <li>EBTKS</li>
    <li>bicgl</li>
    <li>bicpl</li>
    <li>conglomerate</li>
    <li>eccodes</li>
    <li>getdp</li>
    <li>grib-api</li>
    <li>hdf5-mpi</li>
    <li>highfive-mpi</li>
    <li>hpcg</li>
    <li>hpl</li>
    <li>inormalize</li>
    <li>ior</li>
    <li>libminc</li>
    <li>migrate</li>
    <li>minc_tools</li>
    <li>minc_widgets</li>
    <li>mni_autoreg</li>
    <li>mpi (openmpi)</li>
    <li>n3</li>
    <li>netcdf</li>
    <li>netcdfcxx4</li>
    <li>netcdffortran</li>
    <li>oobicpl</li>
    <li>parmetis</li>
    <li>python38Packages.boltztrap2</li>
    <li>python38Packages.h5netcdf</li>
    <li>python38Packages.h5py-mpi</li>
    <li>python38Packages.meshio</li>
    <li>python38Packages.mpi4py</li>
    <li>python38Packages.netcdf4</li>
    <li>python38Packages.wrf-python</li>
    <li>python39Packages.boltztrap2</li>
    <li>python39Packages.eccodes</li>
    <li>python39Packages.h5netcdf</li>
    <li>python39Packages.meshio</li>
    <li>python39Packages.mpi4py</li>
    <li>python39Packages.netcdf4</li>
    <li>python39Packages.wrf-python</li>
    <li>scotch</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Comment by risicle** [comment]
macos 10.15 successfully builds `openmpi` and a small selection of depending packages (probably because nothing's truly changed). Looks useful to me.

File: pkgs/development/libraries/openmpi/default.nix [hunk 1]
```diff
@@ -86,6 +89,7 @@ in stdenv.mkDerivation rec {
 
     sed -i 's:compiler=.*:compiler=${targetPackages.stdenv.cc}/bin/${targetPackages.stdenv.cc.targetPrefix}c++:' \
        $out/share/openmpi/mpic++-wrapper-data.txt
+  '' + lib.optionalString fortranSupport ''
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/openmpi/default.nix:93]:
```suggestion
  '' + lib.optionalString fortranSupport ''
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by purcell** [comment]
>  Is there no option to rather fix gortran on aarch64-darwin? That would be a more natural first step. Lot's of packages need will need the Fortran support.

I've pinged the GCC maintainers on the related issue.

**Comment by purcell** [comment]
It's a bit frustrating to not directly fix the issue, but I haven't been able to get any feedback on its causes, so I feel like this PR might be a reasonable and expedient workaround in the meantime.

**Comment by vcunat** [comment]
> I've pinged the GCC derivation maintainers on the related issue.

I suspect none of them really know/care about fortran...

**Comment by prusnak** [comment]
> I suspect none of them really know/care about fortran...

FWIW this breakage https://github.com/NixOS/nixpkgs/issues/127608#issuecomment-865220282 is not limited to fortran - I have seen it in other aarch64-darwin nixpkgs builds, so it might get fixed eventually by someone fixing the breakage in some other affected package. :-)

**Comment by prusnak** [comment]
Sent another PR with suggestion from vcunat to fix the stackprotector issue: https://github.com/NixOS/nixpkgs/pull/128606

**Pull request merged**